airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/BufferManager.java

```diff
@@ -154,34 +157,47 @@ public Optional<Instant> getTimeOfLastRecord(final StreamDescriptor streamDescri
     public Batch take(final StreamDescriptor streamDescriptor, final long bytesToRead) {
       final var queue = buffers.get(streamDescriptor);
 
-      final AtomicLong bytesRead = new AtomicLong();
-      final var s = Stream.generate(() -> {
-        try {
-          return queue.poll(5, TimeUnit.MILLISECONDS);
-        } catch (final InterruptedException e) {
-          throw new RuntimeException(e);
-        }
-      }).takeWhile(memoryItem -> {
-        // if no new records after waiting, the stream is done.
-        if (memoryItem == null) {
-          return false;
-        }
-
-        // otherwise pull records until we hit the memory limit.
-        final long newSize = memoryItem.size() + bytesRead.get();
-        if (newSize <= bytesToRead) {
-          bytesRead.addAndGet(memoryItem.size());
-          return true;
-        } else {
-          return false;
-        }
-      }).map(MemoryItem::item);
-
-      // todo (cgardens) - possible race where in between pulling records and new records going in that we
-      // reset the limit to be lower than number of bytes already in the queue. probably not a big deal.
-      queue.setMaxMemoryUsage(queue.getMaxMemoryUsage() - bytesRead.get());
+      if (!bufferLocks.containsKey(streamDescriptor)) {
+        bufferLocks.put(streamDescriptor, new ReentrantLock());
+      }
 
-      return new Batch(s, bytesRead.get(), memoryManager);
+      bufferLocks.get(streamDescriptor).lock();
+      try {
+        final AtomicLong bytesRead = new AtomicLong();
+        final List<AirbyteMessage> drainedRecords = new ArrayList<>();
+
+        final var s = Stream.generate(() -> {
+          try {
+            return queue.poll(5, TimeUnit.MILLISECONDS);
+          } catch (final InterruptedException e) {
+            throw new RuntimeException(e);
+          }
+        }).takeWhile(memoryItem -> {
+          // if no new records after waiting, the stream is done.
+          if (memoryItem == null) {
+            return false;
+          }
+
+          // otherwise pull records until we hit the memory limit.
+          final long newSize = memoryItem.size() + bytesRead.get();
+          if (newSize <= bytesToRead) {
+            bytesRead.addAndGet(memoryItem.size());
+            return true;
+          } else {
+            return false;
+          }
+        }).map(MemoryItem::item)
+            .toList()
+            .stream();
```

Comment:
hack for now. we can switch this to a for loop, but didn't want to slow us down on this right now.